vault/resource_pki_secret_backend_key.go

```diff
@@ -0,0 +1,269 @@
+		UpdateContext: pkiSecretBackendKeyUpdate,
+		DeleteContext: pkiSecretBackendKeyDelete,
+		ReadContext:   ReadContextWrapper(ReadContextWrapper(pkiSecretBackendKeyRead)),
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			consts.FieldMount: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Full path where PKI engine is mounted.",
+			},
+			consts.FieldType: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Specifies the type of the key to create.",
+			},
+			consts.FieldKeyName: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "When a new key is created with this request, optionally specifies the name for this.",
+			},
+			consts.FieldKeyType: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				ForceNew:    true,
+				Description: "Specifies the desired key type; must be 'rsa', 'ed25519' or 'ec'.",
+			},
+			consts.FieldKeyBits: {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Computed:    true,
+				ForceNew:    true,
+				Description: "Specifies the number of bits to use for the generated keys.",
+			},
+			consts.FieldKeyID: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "ID of the generated key.",
+			},
+			consts.FieldManagedKeyName: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The managed key's configured name.",
+			},
+			consts.FieldManagedKeyID: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The managed key's UUID.",
+			},
+		},
+	}
+}
+
+func getPKIGenerateKeysPath(mount, keyType string) string {
+	return fmt.Sprintf("%s/keys/generate/%s", mount, keyType)
+}
+
+func getPKIKeysIDPath(mount, keyID string) string {
+	return fmt.Sprintf("%s/key/%s", mount, keyID)
+}
+
+func pkiSecretBackendKeyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	mount := d.Get(consts.FieldMount).(string)
+	keyType := d.Get(consts.FieldType).(string)
+	keyPath := getPKIGenerateKeysPath(mount, keyType)
+
+	fields := []string{
+		consts.FieldKeyName,
+		consts.FieldKeyType,
+		consts.FieldKeyBits,
+	}
+
+	if keyType == kmsType {
+		fields = append(fields, consts.FieldManagedKeyName, consts.FieldManagedKeyID)
+	}
+
+	data := map[string]interface{}{}
+	for _, k := range fields {
+		if v, ok := d.GetOk(k); ok {
+			data[k] = v
+		}
+	}
+
+	resp, err := client.Logical().WriteWithContext(ctx, keyPath, data)
+	if err != nil {
+		return diag.Errorf("error writing data to %q, err=%s", keyPath, err)
+	}
+
+	keyID := resp.Data[consts.FieldKeyID].(string)
```

Comment:
should we check for existence here, in case it's not present in the map?
```
keyID, ok := resp.Data[consts.FieldKeyID]
if !ok {
    ...
}
```

Reply:
Good call! Added in [8b2ccc9](https://github.com/hashicorp/terraform-provider-vault/pull/1829/commits/8b2ccc9b0a39b1e864e554a700ff472abf17a4af)